x/auth/vesting/types/vesting_account.go

```diff
@@ -611,3 +615,290 @@ func marshalYaml(i interface{}) (interface{}, error) {
 	}
 	return string(bz), nil
 }
+
+// Clawback Vesting Account
+
+var _ vestexported.VestingAccount = (*ClawbackVestingAccount)(nil)
+var _ authtypes.GenesisAccount = (*ClawbackVestingAccount)(nil)
+
+// NewClawbackVestingAccount returns a new ClawbackVestingAccount
+func NewClawbackVestingAccount(
+	baseAcc *authtypes.BaseAccount,
+	funder sdk.AccAddress,
+	originalVesting sdk.Coins,
+	startTime int64,
+	lockupPeriods,
+	vestingPeriods Periods,
+) *ClawbackVestingAccount {
+	// copy and align schedules to avoid mutating inputs
+	lp := make(Periods, len(lockupPeriods))
+	copy(lp, lockupPeriods)
+	vp := make(Periods, len(vestingPeriods))
+	copy(vp, vestingPeriods)
+	_, endTime := AlignSchedules(startTime, startTime, lp, vp)
+	baseVestingAcc := &BaseVestingAccount{
+		BaseAccount:     baseAcc,
+		OriginalVesting: originalVesting,
+		EndTime:         endTime,
+	}
+
+	return &ClawbackVestingAccount{
+		BaseVestingAccount: baseVestingAcc,
+		FunderAddress:      funder.String(),
+		StartTime:          startTime,
+		LockupPeriods:      lp,
+		VestingPeriods:     vp,
+	}
+}
+
+// GetVestedCoins returns the total number of vested coins. If no coins are vested,
+// nil is returned.
+func (va ClawbackVestingAccount) GetVestedCoins(blockTime time.Time) sdk.Coins {
+	// It's likely that one or the other schedule will be nearly trivial,
+	// so there should be little overhead in recomputing the conjunction each time.
+	coins := coinsMin(va.GetUnlockedOnly(blockTime), va.GetVestedOnly(blockTime))
+	if coins.IsZero() {
+		return nil
+	}
+	return coins
+}
+
+// GetVestingCoins returns the total number of vesting coins. If no coins are
+// vesting, nil is returned.
+func (va ClawbackVestingAccount) GetVestingCoins(blockTime time.Time) sdk.Coins {
+	return va.OriginalVesting.Sub(va.GetVestedCoins(blockTime))
+}
+
+// LockedCoins returns the set of coins that are not spendable (i.e. locked),
+// defined as the vesting coins that are not delegated.
+func (va ClawbackVestingAccount) LockedCoins(blockTime time.Time) sdk.Coins {
+	return va.BaseVestingAccount.LockedCoinsFromVesting(va.GetVestingCoins(blockTime))
+}
+
+// TrackDelegation tracks a desired delegation amount by setting the appropriate
+// values for the amount of delegated vesting, delegated free, and reducing the
+// overall amount of base coins.
+func (va *ClawbackVestingAccount) TrackDelegation(blockTime time.Time, balance, amount sdk.Coins) {
+	va.BaseVestingAccount.TrackDelegation(balance, va.GetVestingCoins(blockTime), amount)
+}
+
+// GetStartTime returns the time when vesting starts for a periodic vesting
+// account.
+func (va ClawbackVestingAccount) GetStartTime() int64 {
+	return va.StartTime
+}
+
+// GetVestingPeriods returns vesting periods associated with periodic vesting account.
+func (va ClawbackVestingAccount) GetVestingPeriods() Periods {
+	return va.VestingPeriods
+}
+
+// coinEq returns whether two Coins are equal.
+// The IsEqual() method can panic.
+func coinEq(a, b sdk.Coins) bool {
```

Comment:
We should make this public as we execute the exact same  check in the message handler I noticed.